src/main/java/com/snowflake/kafka/connector/internal/streaming/SnowflakeSinkServiceV2.java

```diff
@@ -349,15 +361,40 @@ public void callAllGetOffset() {
 
   @Override
   public void closeAll() {
+    if (closeChannelsInParallel) {
+      closeAllInParallel();
+    } else {
+      closeAllSequentially();
+    }
+
+    partitionsToChannel.clear();
+
+    StreamingClientProvider.getStreamingClientProviderInstance()
+        .closeClient(this.connectorConfig, this.streamingIngestClient);
+  }
+
+  private void closeAllSequentially() {
     partitionsToChannel.forEach(
         (partitionChannelKey, topicPartitionChannel) -> {
           LOGGER.info("Closing partition channel:{}", partitionChannelKey);
           topicPartitionChannel.closeChannel();
         });
-    partitionsToChannel.clear();
+  }
 
-    StreamingClientProvider.getStreamingClientProviderInstance()
-        .closeClient(this.connectorConfig, this.streamingIngestClient);
+  private void closeAllInParallel() {
```

Comment:
Think about more I'm not sure this approach would help a lot. If you look into the close function in the SDK, it will flush everything in the client and then call get status on this particular channel. If we're doing it in parallel, the same number of flush will be called, every flush will still be run in serial and block other flushes. I thought most of time is spent on waiting for server side to commit, which will stay the same since everything will be flush at once. Have you verified locally that it reduces the close time dramatically with a large number of channels?

Reply:
Hi @sfc-gh-tzhang, as Lex is OOO, I prepared a short test for that. I am not sure whether the IT is sufficient to give meaningful information, but I ran [such a test](https://github.com/snowflakedb/snowflake-kafka-connector/commit/7d6a84236a7ba7522e3a713e2fe5e61a151f20fb#diff-4005e4dba838ef1cd8e0fe2c4b34e061ed69ec1da322bd1ebf1503e8be051a1eR225-R269) closing 50 channels. It shows the parallelism speeded the execution up.

13513.7 ms mean of sequential execution time vs 1508.0 of parallel
<img width="566" alt="image" src="https://github.com/snowflakedb/snowflake-kafka-connector/assets/85181931/071f3947-6a2b-42b1-8253-119d232a5806">

But the scenario does not cover the case where we have rows present in `net.snowflake.ingest.streaming.internal.SnowflakeStreamingIngestChannelInternal` buffer (the scenario described by you) - I cannot reproduce it even though I played with the buffering parameters.

Reply:
The log level was set to DEBUG so it added a few additional seconds to the measured time. I changed it to INFO and attached the profiler. It takes ~11 seconds and spends most of its time on `CompletableFuture.get()`. 
<img width="1201" alt="image" src="https://github.com/snowflakedb/snowflake-kafka-connector/assets/85181931/b2fd6d1f-4992-45a1-9f4e-b852337904d4">